decoder/signature.go

```diff
@@ -0,0 +1,123 @@
+package decoder
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/hashicorp/hcl-lang/lang"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+)
+
+// SignatureAtPos returns a function signature for the given pos if pos
+// is inside a FunctionCallExpr
+func (d *PathDecoder) SignatureAtPos(filename string, pos hcl.Pos) (*lang.FunctionSignature, error) {
+	file, err := d.fileByName(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := d.bodyForFileAndPos(filename, file, pos)
+	if err != nil {
+		return nil, err
+	}
+
+	var signature *lang.FunctionSignature
+	hclsyntax.VisitAll(body, func(node hclsyntax.Node) hcl.Diagnostics {
+		if !node.Range().ContainsPos(pos) {
+			return nil // Node not under cursor
+		}
```

Comment:
With the current implementation, the feature is decoupled from any expression work and can work stand-alone. Also, nested expressions/function calls should work out of the box, as we always try to find the specific `FunctionCallExpr` node.

There is no overlap between the functions in the schema and the type declarations. So, even though `list(` or `tuple(` is technically a `FunctionCallExpr`, no signature help should come up. Only for Terraform `< 0.15` do we have the deprecated `list` and `map` functions which will trigger signature help. Is it worth extending the code to account for that? Adding schema and constraint checks while walking the AST sounds expensive.

Reply:
I wouldn't think it's expensive in terms of performance (not significantly more than the implementation of `hclsyntax.VisitAll`). I can totally see it being expensive in terms of effort/complexity though and I can see why this is worth shipping as-is and how it can deliver immense value even if we don't account for these edge cases. 👍🏻 

So with that in mind I'd just leave a note either in a comment, or in a new issue to help us remember to revisit this and/or explain why we took a slightly different approach here which doesn't take schema into account and call the PR done.